File: tradebot/options.py
"""Black–Scholes call option pricing + greeks.

Only European calls are modelled, which is what we trade (buy then sell before
expiration, or let expire worthless). All math is closed-form; no numerical
integration needed for the MVP.
"""
from __future__ import annotations

import math
from dataclasses import dataclass


SQRT_2PI = math.sqrt(2.0 * math.pi)


def _norm_cdf(x: float) -> float:
    return 0.5 * (1.0 + math.erf(x / math.sqrt(2.0)))


def _norm_pdf(x: float) -> float:
    return math.exp(-0.5 * x * x) / SQRT_2PI


@dataclass
class CallQuote:
    price: float
    delta: float
    gamma: float
    theta: float   # per year; divide by 365 for daily
    vega: float    # per 1.00 change in sigma
    d1: float
    d2: float


def price_call(
    s: float,   # spot
    k: float,   # strike
    t: float,   # time to expiration in YEARS
    r: float,   # risk-free rate
    sigma: float,
) -> CallQuote:
    """Return the Black–Scholes price and greeks for a European call.

    Handles the ``t == 0`` edge case (expiration day) by returning intrinsic value.
    """
    if t <= 0 or sigma <= 0:
        intrinsic = max(s - k, 0.0)
        delta = 1.0 if s > k else 0.0
        return CallQuote(price=intrinsic, delta=delta, gamma=0.0, theta=0.0,
                         vega=0.0, d1=float("nan"), d2=float("nan"))

    sqrt_t = math.sqrt(t)
    d1 = (math.log(s / k) + (r + 0.5 * sigma * sigma) * t) / (sigma * sqrt_t)
    d2 = d1 - sigma * sqrt_t
    nd1 = _norm_cdf(d1)
    nd2 = _norm_cdf(d2)
    pdf1 = _norm_pdf(d1)

    price = s * nd1 - k * math.exp(-r * t) * nd2
    delta = nd1
    gamma = pdf1 / (s * sigma * sqrt_t)
    theta = (-(s * pdf1 * sigma) / (2 * sqrt_t)
             - r * k * math.exp(-r * t) * nd2)
    vega = s * pdf1 * sqrt_t
    return CallQuote(price=price, delta=delta, gamma=gamma, theta=theta,
                     vega=vega, d1=d1, d2=d2)


def choose_strike(spot: float, offset_pct: float) -> float:
    """Round strike to nearest $1 increment, offset_pct OTM (positive = above spot)."""
    raw = spot * (1.0 + offset_pct)
    return round(raw)

File: lib/options.ts
// Black-Scholes call pricer + greeks. Pure math, no deps.

export interface CallQuote {
  price: number;
  delta: number;
  gamma: number;
  theta: number; // per year
  vega: number;
}

const SQRT_2PI = Math.sqrt(2 * Math.PI);

function erf(x: number): number {
  // Abramowitz & Stegun 7.1.26 approximation. Accurate to ~1e-7.
  const sign = x < 0 ? -1 : 1;
  x = Math.abs(x);
  const a1 = 0.254829592;
  const a2 = -0.284496736;
  const a3 = 1.421413741;
  const a4 = -1.453152027;
  const a5 = 1.061405429;
  const p = 0.3275911;
  const t = 1.0 / (1.0 + p * x);
  const y =
    1.0 -
    ((((a5 * t + a4) * t + a3) * t + a2) * t + a1) * t * Math.exp(-x * x);
  return sign * y;
}

export function normCdf(x: number): number {
  return 0.5 * (1 + erf(x / Math.SQRT2));
}

export function normPdf(x: number): number {
  return Math.exp(-0.5 * x * x) / SQRT_2PI;
}

/** Price a European call. `t` is years to expiration. */
export function priceCall(
  s: number,
  k: number,
  t: number,
  r: number,
  sigma: number,
): CallQuote {
  if (t <= 0 || sigma <= 0) {
    const intrinsic = Math.max(s - k, 0);
    return {
      price: intrinsic,
      delta: s > k ? 1 : 0,
      gamma: 0,
      theta: 0,
      vega: 0,
    };
  }
  const sqrtT = Math.sqrt(t);
  const d1 = (Math.log(s / k) + (r + 0.5 * sigma * sigma) * t) / (sigma * sqrtT);
  const d2 = d1 - sigma * sqrtT;
  const nd1 = normCdf(d1);
  const nd2 = normCdf(d2);
  const pdf1 = normPdf(d1);
  const price = s * nd1 - k * Math.exp(-r * t) * nd2;
  return {
    price,
    delta: nd1,
    gamma: pdf1 / (s * sigma * sqrtT),
    theta:
      -(s * pdf1 * sigma) / (2 * sqrtT) - r * k * Math.exp(-r * t) * nd2,
    vega: s * pdf1 * sqrtT,
  };
}

/** Nearest-dollar strike offset_pct above spot (positive = OTM). */
export function chooseStrike(spot: number, offsetPct: number): number {
  return Math.round(spot * (1 + offsetPct));
}
